2.mutation-classifier_(WIP_Gene-Weights).py

```diff
@@ -0,0 +1,439 @@
+
+metrics_dict = {    
+    model: {
+        'train': get_threshold_metrics(y_train, y_pred_dict[model]['train']),
+        'test':  get_threshold_metrics(y_test, y_pred_dict[model]['test'])
+    } for model in y_pred_dict.keys()
+}
+
+
+# In[18]:
+
+# Assemble the data for ROC curves
+model_order = ['full', 'expressions', 'covariates']
+
+auc_output = pd.DataFrame()
+roc_output = pd.DataFrame()
+
+for model in model_order:
+    metrics_partition = metrics_dict[model]
+    for partition, metrics in metrics_partition.items():
+        auc_output = auc_output.append(pd.DataFrame({
+            'partition': [partition],
+            'feature_set': [model],
+            'auc': metrics['auroc']
+        }))
+        roc_df = metrics['roc_df']
+        roc_output = roc_output.append(pd.DataFrame({
+            'false_positive_rate': roc_df.fpr,
+            'true_positive_rate': roc_df.tpr,
+            'partition': partition,
+            'feature_set': model
+        }))
+auc_output['legend_index'] = range(len(auc_output.index))
+
+with open('vega_specs/roc_vega_spec.json', 'r') as fp:
+    vega_spec = json.load(fp)
+
+final_spec = fill_spec_with_data(vega_spec, 
+    {'roc': roc_output, 'legend_auc': auc_output})
+
+Vega(final_spec)
+
+
+# ## What are the classifier coefficients?
+
+# In[19]:
+
+final_pipelines = {
+    model: pipeline.best_estimator_
+    for model, pipeline in cv_pipelines.items()
+}
+final_classifiers = {
+    model: pipeline.named_steps['classify']
+    for model, pipeline in final_pipelines.items()
+}
+
+coef_df = pd.concat([
+    get_model_coefficients(classifier, model, covariate_df.columns)
+    for model, classifier in final_classifiers.items()
+])
+
+
+# In[20]:
+
+# Signs of the coefficients by model
+pd.crosstab(coef_df.feature_set, np.sign(coef_df.weight).rename('coefficient_sign'))
+
+
+# In[21]:
+
+# Top standardized coefficients for the full data-set
+(coef_df
+    .query("feature_set == 'full'")
+    .head(10)
+)
+
+
+# # Attempt to Identify top genes
+
+# In[22]:
+
+# Top standardized coefficients for the expression data-set
+(coef_df
+    .query("feature_set == 'expressions'")
+    .head(10)
+)
+
+
+# ## Becuase the above features are PCA components they don't provide any useful information about which genes were used/most impactful in the classifier. The three cells below is my initial attempt to find out which _genes_ were most impactful.
+
+# Create a dataframe with coefficients from the classifier for the expression only data-set
+
+# In[23]:
+
+coefficients = coef_df.query("feature_set == 'expressions'")['weight']
+coefficients = coefficients.to_frame()
+print(coefficients.shape)
+print(coefficients.head(5))
+
+
+# Create a dataframe with the weights used in the PCA transform
+
+# In[24]:
+
+# Run PCA on the X_train expression data
+n_genes = len(expression_df.columns)
+X_train_expression = X_train[X.columns[-n_genes:]]
+pca = PCA(n_components = 50, random_state = 0)
+pca.fit(X_train_expression)
```

Comment:
Hmm you're refitting the PCA? Shouldn't we extract the `components_` from the transformation already fit in the pipeline rather than create a new transformation?